File: src/main/java/com/lmp/catalog/dto/ServiceResponse.java
package com.lmp.catalog.dto;

import com.lmp.catalog.domain.ServiceBenefit;
import com.lmp.catalog.domain.ServiceOffer;
import com.lmp.shared.pricing.PricingContext;
import com.lmp.shared.pricing.RegionalPricingService;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * DTO de réponse pour un service du catalogue.
 */
public record ServiceResponse(
        UUID id,
        String title,
        String slug,
        String description,
        String icon,
        String categoryName,
        String categorySlug,
        boolean featured,
        boolean active,
        int displayOrder,
        List<String> benefits,
        OfferResponse currentOffer
) {
    /** Catalogue admin / sans conversion régionale (prix EUR en base). */
    public static ServiceResponse from(com.lmp.catalog.domain.Service service) {
        List<String> benefitList = service.getBenefits() != null
                ? service.getBenefits().stream().map(ServiceBenefit::getBenefit).collect(Collectors.toList())
                : List.of();

        ServiceOffer offer = service.getCurrentOffer();
        OfferResponse offerResponse = offer != null ? OfferResponse.from(offer) : null;

        return new ServiceResponse(
                service.getId(),
                service.getTitle(),
                service.getSlug(),
                service.getDescription(),
                service.getIcon(),
                service.getCategory() != null ? service.getCategory().getName() : null,
                service.getCategory() != null ? service.getCategory().getSlug() : null,
                Boolean.TRUE.equals(service.getFeatured()),
                Boolean.TRUE.equals(service.getActive()),
                service.getDisplayOrder() != null ? service.getDisplayOrder() : 0,
                benefitList,
                offerResponse);
    }

    /** API publique : prix convertis selon l'IP du client. */
    public static ServiceResponse from(com.lmp.catalog.domain.Service service,
            PricingContext ctx,
            RegionalPricingService pricing) {
        List<String> benefitList = service.getBenefits() != null
                ? service.getBenefits().stream().map(ServiceBenefit::getBenefit).collect(Collectors.toList())
                : List.of();

        ServiceOffer offer = service.getCurrentOffer();
        OfferResponse offerResponse = offer != null ? OfferResponse.from(offer, ctx, pricing) : null;

        return new ServiceResponse(
                service.getId(),
                service.getTitle(),
                service.getSlug(),
                service.getDescription(),
                service.getIcon(),
                service.getCategory() != null ? service.getCategory().getName() : null,
                service.getCategory() != null ? service.getCategory().getSlug() : null,
                Boolean.TRUE.equals(service.getFeatured()),
                Boolean.TRUE.equals(service.getActive()),
                service.getDisplayOrder() != null ? service.getDisplayOrder() : 0,
                benefitList,
                offerResponse);
    }

    public record OfferResponse(
            UUID id,
            String name,
            BigDecimal price,
            BigDecimal originalPrice,
            String durationType,
            boolean isDefault,
            String currency,
            BigDecimal priceEur
    ) {
        public static OfferResponse from(ServiceOffer offer) {
            BigDecimal p = offer.getPrice();
            return new OfferResponse(
                    offer.getId(),
                    offer.getName(),
                    p,
                    offer.getOriginalPrice(),
                    offer.getDurationType() != null ? offer.getDurationType().name() : null,
                    offer.getIsDefault(),
                    "EUR",
                    p);
        }

        public static OfferResponse from(ServiceOffer offer, PricingContext ctx, RegionalPricingService pricing) {
            BigDecimal eur = offer.getPrice();
            BigDecimal eurOrig = offer.getOriginalPrice();
            BigDecimal display = pricing.convertFromEur(eur, ctx);
            BigDecimal displayOrig = eurOrig != null ? pricing.convertFromEur(eurOrig, ctx) : null;
            return new OfferResponse(
                    offer.getId(),
                    offer.getName(),
                    display,
                    displayOrig,
                    offer.getDurationType() != null ? offer.getDurationType().name() : null,
                    offer.getIsDefault(),
                    ctx.currency(),
                    eur);
        }
    }
}

File: backend-dotnet/src/Lmp.Application/Catalog/ServiceResponse.cs
using Lmp.Application.Pricing;
using Lmp.Domain.Catalog;

namespace Lmp.Application.Catalog;

/// <summary>
/// Catalogue service response DTO. Faithful port of
/// <c>com.lmp.catalog.dto.ServiceResponse</c>.
/// </summary>
public sealed record ServiceResponse(
    Guid Id,
    string Title,
    string Slug,
    string Description,
    string? Icon,
    string? CategoryName,
    string? CategorySlug,
    bool Featured,
    bool Active,
    int DisplayOrder,
    IReadOnlyList<string> Benefits,
    ServiceResponse.OfferResponse? CurrentOffer)
{
    /// <summary>Admin catalogue / no regional conversion (EUR prices from the database).</summary>
    public static ServiceResponse From(Service service)
    {
        var benefits = service.Benefits.Select(b => b.Benefit).ToList();
        var offer = service.GetCurrentOffer();
        return Build(service, benefits, offer is null ? null : OfferResponse.From(offer));
    }

    /// <summary>Public API: prices converted to the client's resolved currency.</summary>
    public static ServiceResponse From(Service service, PricingContext context, IRegionalPricingService pricing)
    {
        var benefits = service.Benefits.Select(b => b.Benefit).ToList();
        var offer = service.GetCurrentOffer();
        return Build(service, benefits, offer is null ? null : OfferResponse.From(offer, context, pricing));
    }

    private static ServiceResponse Build(Service service, IReadOnlyList<string> benefits, OfferResponse? offer)
        => new(
            service.Id,
            service.Title,
            service.Slug,
            service.Description,
            service.Icon,
            service.Category?.Name,
            service.Category?.Slug,
            service.Featured,
            service.Active,
            service.DisplayOrder,
            benefits,
            offer);

    public sealed record OfferResponse(
        Guid Id,
        string Name,
        decimal? Price,
        decimal? OriginalPrice,
        string? DurationType,
        bool IsDefault,
        string Currency,
        decimal? PriceEur)
    {
        public static OfferResponse From(ServiceOffer offer)
            => new(
                offer.Id,
                offer.Name,
                offer.Price,
                offer.OriginalPrice,
                DurationTypeWire(offer.DurationType),
                offer.IsDefault,
                "EUR",
                offer.Price);

        public static OfferResponse From(ServiceOffer offer, PricingContext context, IRegionalPricingService pricing)
        {
            var display = pricing.ConvertFromEur(offer.Price, context);
            var displayOriginal = offer.OriginalPrice is null
                ? null
                : pricing.ConvertFromEur(offer.OriginalPrice, context);
            return new OfferResponse(
                offer.Id,
                offer.Name,
                display,
                displayOriginal,
                DurationTypeWire(offer.DurationType),
                offer.IsDefault,
                context.Currency,
                offer.Price);
        }
    }

    /// <summary>Wire name matching the Java enum (<c>DurationType.name()</c>).</summary>
    private static string DurationTypeWire(DurationType type) => type switch
    {
        Lmp.Domain.Catalog.DurationType.OneTime => "ONE_TIME",
        Lmp.Domain.Catalog.DurationType.Monthly => "MONTHLY",
        Lmp.Domain.Catalog.DurationType.Yearly => "YEARLY",
        _ => type.ToString(),
    };
}
